nomad/task_group_host_volume_claim_endpoint.go

```diff
@@ -0,0 +1,159 @@
+// Copyright (c) HashiCorp, Inc.
+// SPDX-License-Identifier: BUSL-1.1
+
+package nomad
+
+import (
+	"fmt"
+	"net/http"
+	"strings"
+	"time"
+
+	"github.com/hashicorp/go-hclog"
+	"github.com/hashicorp/go-memdb"
+	metrics "github.com/hashicorp/go-metrics/compat"
+	"github.com/hashicorp/nomad/acl"
+	"github.com/hashicorp/nomad/nomad/state"
+	"github.com/hashicorp/nomad/nomad/state/paginator"
+	"github.com/hashicorp/nomad/nomad/structs"
+)
+
+// TaskGroupHostVolumeClaim is the server RPC endpoint for task group volume claims
+type TaskGroupHostVolumeClaim struct {
+	srv    *Server
+	ctx    *RPCContext
+	logger hclog.Logger
+}
+
+func NewTaskGroupVolumeClaimEndpoint(srv *Server, ctx *RPCContext) *TaskGroupHostVolumeClaim {
+	return &TaskGroupHostVolumeClaim{srv: srv, ctx: ctx, logger: srv.logger.Named("task_group_host_volume_claim")}
+}
+
+func (tgvc *TaskGroupHostVolumeClaim) List(args *structs.TaskGroupVolumeClaimListRequest, reply *structs.TaskGroupVolumeClaimListResponse) error {
+	authErr := tgvc.srv.Authenticate(tgvc.ctx, args)
+	if done, err := tgvc.srv.forward(structs.TaskGroupHostVolumeClaimListRPCMethod, args, args, reply); done {
+		return err
+	}
+	tgvc.srv.MeasureRPCRate("task_group_volume_claim", structs.RateMetricList, args)
+	if authErr != nil {
+		return structs.ErrPermissionDenied
+	}
+	defer metrics.MeasureSince([]string{"nomad", "task_group_volume_claim", "list"}, time.Now())
+
+	allowClaim := acl.NamespaceValidator(acl.NamespaceCapabilityHostVolumeRead)
+	aclObj, err := tgvc.srv.ResolveACL(args)
+	if err != nil {
+		return err
+	}
+	if !allowClaim(aclObj, args.RequestNamespace()) {
+		return structs.ErrPermissionDenied
+	}
+
+	ns := args.RequestNamespace()
+
+	opts := blockingOptions{
+		queryOpts: &args.QueryOptions,
+		queryMeta: &reply.QueryMeta,
+		run: func(ws memdb.WatchSet, stateStore *state.StateStore) error {
+			iter, err := stateStore.TaskGroupHostVolumeClaimsByFields(ws,
+				state.TgvcSearchableFields{
+					Namespace:     ns,
+					JobID:         args.JobID,
+					VolumeName:    args.VolumeName,
+					TaskGroupName: args.TaskGroup,
+				})
+			if err != nil {
+				return err
+			}
+
+			tokenizer := paginator.NewStructsTokenizer(iter,
+				paginator.StructsTokenizerOptions{
+					WithNamespace: true,
+					WithID:        true,
+				},
+			)
+
+			filters := []paginator.Filter{
+				paginator.GenericFilter{
+					Allow: func(raw any) (bool, error) {
+						claim := raw.(*structs.TaskGroupHostVolumeClaim)
+						// empty prefix doesn't filter
+						if !strings.HasPrefix(claim.ID, args.Prefix) {
+							return false, nil
+						}
+
+						allowClaim := acl.NamespaceValidator(acl.NamespaceCapabilityHostVolumeRead)
```

Comment:
You can move this line before `filters := []paginator.Filter` so we don't have to call this method to get a function pointer for each item we iterate.